### ISSUE TITLE
Performance bottleneck

### ISSUE THREAD

**Issue by grayfall** [Illumina/SpliceAI#48]
Greetings. 

https://github.com/Illumina/SpliceAI/blob/76ef672108c429cac1f49226ef87bdf8d9932482/spliceai/utils.py#L158

Have you considered, that this line (along with the next one) is a huge performance bottleneck, because you basically calculate a small batch whilst loading/unloading models from GPU memory? GPUs are only effective when you load them consistently with large batches without any context-switches.

**Comment by kishorejaganathan** [comment]
We did look into it. The GPU utilization while annotating variants for the current code is around 65-70%, so any further optimization is only going to increase the speed by 30-40% but the code becomes a lot less modular. This is particularly an issue when there are INDELs in the inputs (Keras requires the entire batch to have the same size), so we decided to not optimize this.

**Comment by grayfall** [comment]
Thanks for the feedback. A significant part of those 65-70% might be due to loading/unloading the models, not actual computations.

**Comment by grayfall** [comment]
I've been trying to reimplement the `get_delta_scores` function to separate the logic that creates data for predictions, does the predictions and does post-processing, but it's very hard to follow the execution flow without any comments.

**Comment by kishorejaganathan** [comment]
Just in case you are doing this because you have to annotate a large file, I'd like to point out that we have already pre-annotated all possible SNVs and short INDELs within genes (please see README for the file link).

**Comment by grayfall** [comment]
@kishorejaganathan we want to generate a similar annotation database for an alternative genome assembly, which is why we are looking into potential optimisations.

**Comment by grayfall** [comment]
@kishorejaganathan we've been able to achieve a 3x performance increase by fixing the issues with this bottleneck in a synthetic test. We are now definitely going to reimplement `get_delta_scores`. I suspect you might be short on time, but is there any chance you could add some comments to `get_delta_scores` to make the data-preprocessing and post-processing steps more clear for an outside reader?

**Comment by kishorejaganathan** [comment]
That is a pretty healthy performance increase. Everything below the lines you are interested in (158-159) is post-processing, and everything before is pre-processing. The lines inside the defined class process the input annotation file and implement some basic functionality like finding which gene the variant lies in, where is the closest exon, etc. The lines in get_delta_scores function before the lines you are interested in just filter out variants that are not compatible with SpliceAI, and prepare the input for the variants that are compatible...

**Comment by grayfall** [comment]
@kishorejaganathan 
Hi. Can you explain why you add 1 to exon starts here:
https://github.com/Illumina/SpliceAI/blob/76ef672108c429cac1f49226ef87bdf8d9932482/spliceai/utils.py#L25

**Comment by kishorejaganathan** [comment]
SpliceAI gives a score to each nucleotide in the input sequence, the convention we used while training SpliceAI was to assign high acceptor scores to the first nucleotide of each exon and high donor scores to the last nucleotide of each exon. The GENCODE annotation files provide the last nucleotide of each intron/exon, the +1 moves the nucleotide of interest from the last of the intron to the first of the exon.

**Comment by grayfall** [comment]
@kishorejaganathan Thanks for the quick response. Do I get it right: GENCODE always gives you EXON_START - 1 instead of the actual EXON_START?

**Comment by kishorejaganathan** [comment]
That is correct, you can verify this by viewing a couple of positions in the UCSC genome browser.

**Comment by grayfall** [comment]
@kishorejaganathan One more question. Whilst training the model, did you train on individual variants (as you do prediction) or did you join multiple records and trained on batches? I'm asking this because we are thinking whether joining encoding in batches will affect predictions. References sequences always have the same length, but alternative sequences have variable length, forcing us to add additional padding to make all records in a batch congruent.

**Comment by kishorejaganathan** [comment]
Training was done purely using reference sequences and no variant data was used (i.e., input is the reference sequence and output for each position is whether that position is used as a splice site or not). We used batches while training, unlike the code we've provided for variant effect prediction.

**Comment by grayfall** [comment]
@kishorejaganathan I ended up creating batches of sequences with identical length to overcome the aforementioned issue. I've finished reimplementing SpliceAI. I used the bundled input.vcf duplicated 200 times to compare performance. On an RTX 2080 I get 3.5 times performance increase relative to the original implementation. Some of it comes from minor improvements in preprocessing routines. I still need to write a `setup.py` script, but that's a minor detail. https://github.com/skoblov-lab/spliceai-reforged
